Add Inmar bidder adapter

zandree-owneriq opened the pull request:
## Type of change
- [x] New bidder adapter

## Description of change
Add Inmar bidder adapter

- test parameters for validating bids
```
{
  bidder: 'inmar',
  params: {
    partnerId: 13144370,
    adnetId: 'ADb1f40rmo',
    bidFloor: 0.40,
    position: 1
  }
}
```

Be sure to test the integration with your adserver using the [Hello World](/integrationExamples/gpt/hello_world.html) sample page.

- contact email of the adapter’s maintainer
oiq_rtb@inmar.com
- [x] official adapter submission

For any changes that affect user-facing APIs or example code documented on http://prebid.org, please provide:

- A link to a PR on the docs repo at https://github.com/prebid/prebid.github.io/
https://github.com/prebid/prebid.github.io/pull/2118

## Other information
Wasn't able to reopen https://github.com/prebid/Prebid.js/pull/5467 so I'm creating a new pull request

patmmccann commented on test/spec/modules/inmarBidAdapter_spec.js:
```diff
@@ -0,0 +1,257 @@
+    body: {
+      cpm: 1.50,
+      ad: '<!-- script -->',
+      mediaType: 'js',
```
banner?

patmmccann commented on modules/inmarBidAdapter.js:
```diff
@@ -0,0 +1,115 @@
+      gdprConsent: bidderRequest.gdprConsent,
+      uspConsent: bidderRequest.uspConsent,
+      currencyCode: config.getConfig('currency.adServerCurrency'),
+      domain: config.getConfig('publisherDomain'),
```
getting deprecated https://github.com/prebid/Prebid.js/pull/5674/files

patmmccann replied:
https://github.com/prebid/Prebid.js/issues/5637 rather

patmmccann commented on modules/inmarBidAdapter.md:
```diff
@@ -0,0 +1,48 @@
+                        params: {
+                            partnerId: 12345,
+                            adnetId: 'ADb1f40rmi',
+                            bidFloor: 0.70,
```
you must use add support for the floors module if you are reading floors

patmmccann commented on modules/inmarBidAdapter.js:
```diff
@@ -0,0 +1,115 @@
+          width: response.width,
+          height: response.height,
+          ad: response.ad,
+          ttl: response.ttl,
```
please consider adding meta fields as per #3115

zandree-owneriq commented:
1) What is getting deprecated?  There are four lines of code that show up and I don't know which one you're referring to.

2) Is there a more detailed spec for the meta block beyond this proposal thread?  This is the first I've heard of it.

3) CircleCI failed the unit tests, but these failures appear to be outside of my code.  Is there something I need to fix, or should I ignore it?

patmmccann commented:
1. publisher domain is getting deprecated; see https://github.com/prebid/Prebid.js/issues/5637

2. it is in the how to add a bid adapter dev doc now; search for meta

3. I'm afraid I am not sure; rebasing to master will restart circleci

zandree-owneriq commented:
Ok, I've removed publisherDomain and bidFloor, fixed the unit tests, and added the meta block, which will be filled out on our servers with as much info as we can.  I've pushed the changes to my forked repo - do I need to do anything to update this pull request, or does that happen automatically?

patmmccann commented:
I can see all your changes, but why remove support for floors instead of adding support for the getFloor function?

zandree-owneriq commented:
That decision was above me.  We have some time pressure to get this working on a basic level.  We may add support for floors in a future release.

Is there anything else I need to do to get this merged?

zandree-owneriq commented:
Any word on this?

patmmccann commented:
@zandree-owneriq any luck on getting your circle ci setup properly?

zandree-owneriq commented:
No, when I asked before about whether I needed to fix something or if I could ignore it (since the errors were outside of my adapter, and I didn't change any other code), you said you weren't sure.  And every time I asked what I needed to get my code merged, CircleCI wasn't mentioned, so I figured it wasn't a roadblock to getting merged.

If there is something on my end that is preventing my adapter from being merged, please let me know what I need to do.

patmmccann commented:
You appear to have circleci configured for your own organization and
the wrong tests are getting published.

On Tue, Sep 8, 2020 at 7:14 PM Zak Andree <notifications@github.com> wrote:
>
> No, when I asked before about whether I needed to fix something or if I could ignore it (since the errors were outside of my adapter, and I didn't change any other code), you said you weren't sure. And every time I asked what I needed to get my code merged, CircleCI wasn't mentioned, so I figured it wasn't a roadblock to getting merged.
>
> If there is something on my end that is preventing my adapter from being merged, please let me know what I need to do.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.

zandree-owneriq commented:
My copy of .circleci/config.yml is identical to the one in the upstream repository.  Where do you see that I have it configured for my own organization?

patmmccann commented:
From the existence of the e2etest that is on no other pull requests and the build tests failing to authorize

zandree-owneriq commented:
e2etest exists in the config file on the main Prebid.js repo as well:
https://github.com/prebid/Prebid.js/blob/master/.circleci/config.yml#L76

To my knowledge, we have never used CircleCI at Inmar, and I did nothing to change the CircleCI setup.  Again, my .circleci/config.yml is identical to the one in the upstream branch.  If you have any insight into what is making it behave differently for me, or more importantly how I could fix it, that would be greatly appreciated.

zandree-owneriq commented:
Merging the upstream master into my repo appears to have solved the e2etest issue, but now the build test won't even run.  The error message is: "This job has been blocked because resource-class:xlarge is not available on your plan. Please upgrade to continue building."  It appears a change was made in .circleci/config.yml earlier today to add "resource_class: xlarge".  You guys may want to roll that back if it's not available to everyone.

zandree-owneriq commented:
Is there any update here?

zandree-owneriq commented:
Just following up again.  Is there any update on this?  Wondering if the comment from @jsnellbaker in that pull request from last November made any sense.  If we can't figure this out, would it be possible to get on a call with someone?  Thanks.

patmmccann commented:
I'm trying to identify someone that can help you; I am not familiar enough with circleci to troubleshoot your issues.

harpere commented:
@zandree-owneriq @patmmccann  this comment may help:  https://github.com/prebid/Prebid.js/pull/4740#issuecomment-576703428

zandree-owneriq commented:
Looks like unfollowing our forked project in CircleCI did the trick.  Ok to merge now?

mike-chowla commented on modules/inmarBidAdapter.js:
```diff
@@ -0,0 +1,113 @@
+      method: 'POST',
+      url: 'https://prebid.owneriq.net:8443/bidder/pb/bid',
+      options: {
+        withCredentials: false
```
Confused as to why you are sending withCredentials as false since that makes user sync pointless since you wont get your cookie back on the bid request

zandree-owneriq commented:
@mike-chowla When using the hello_world.html test, we got the following error message:

Access to XMLHttpRequest at 'https://prebid-test.owneriq.net/bidder/pb/bid' from origin 'https://sandbox.owneriq.net' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.

So we added Access-Control-Allow-Origin * to the response headers, and got this error:

Access to XMLHttpRequest at 'https://prebid-test.owneriq.net/bidder/pb/bid' from origin 'https://sandbox.owneriq.net' has been blocked by CORS policy: The value of the 'Access-Control-Allow-Origin' header in the response must not be the wildcard '*' when the request's credentials mode is 'include'. The credentials mode of requests initiated by the XMLHttpRequest is controlled by the withCredentials attribute.

Changing withCredentials to false removed this error.  If there is a better way to solve this, please let me know.

gloeffelholz commented:
Hello @mike-chowla or anyone else that can help - I am working with @zandree-owneriq to finalize our final submission for our prebid adapter.  Can you or someone else please confirm 2 things:

1. It sounds like "withCredentials" should be true so we get our User IDs that get synced back in bid requests.  Are you familiar with the issue brought up by Zak above?  Is that something only related to our testing that will not be an issue once we go live?  I found more info here, but not sure how relevant it is: https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS/Errors/CORSMissingAllowOrigin.  We absolutely want to get our UserID back in bid requests.  

2. That leads me to the 2nd question - where exactly will our User ID show up in the bid requests after the 1st bid request where we only get ID from the publisher and/or supported User ID sub-modules?   It seems like our User ID (cookie ID) would show up in "bidRequest.userID".  Is that correct?  How do we confirm the exact name of our ID?  Is that something we are dictating in our final Prebid.js documentation?   It seems like whatever we determine would end up showing up publicly on this page for publishers to see: https://docs.prebid.org/dev-docs/modules/userId.html#prebidjs-adapters. 

Is that correct?  If not, please help fill the gaps.  If we're at least close how are we setting these Fields needed:

- ID System Name
- ID System Host
- Prebid.js Attr
- Example Value

bretg commented:
docs PR https://github.com/prebid/prebid.github.io/pull/2585

gloeffelholz commented:
Do I need to post in a public forum to get answers to the above questions?

The documentation on Prebid really is incomplete around where our Cookie ID will be sent back to us as a result of the GetUserSync process.  At a minimum we need someone to clarify that.

bretg commented:
@gloeffelholz and @zandree-owneriq - I suspect that your endpoint needs to dynamically specify the response value of Access-Control-Allow-Origin to the exact origin specified on the request. Wildcard isn't going to work. @harpere or @jsnellbaker: please confirm? Looking through the other bid adapters, it's fairly evenly split between those that specify withCredentials: true vs false. Do we have technical guidance here?

> where exactly will our User ID show up in the bid requests after the 1st bid request 

Your server will see it on the HTTP 'cookie' header.

Bidder-specific user syncs are just standard 3rd party cookies, not part of the PBJS User ID module. That module is reserved for "global" (cross-bidder) IDs.

gloeffelholz commented:
@bretg - We are just releasing an adapter for now, not a prebid server.  So both product and engineering on our side are still not understanding how our standard 3rd party cookie is going to show up in the bid requests that are sent to us.  The endpoint we provided will just give the publisher our Cookie ID.  I understand that will store in 3rd party cookie space.  However, when our adapater is placed by a publisher we should then receive bid requests.  The publisher initiates those through GAM for example and our adapter receives the request.  We are not clear on how we will be able to see the "http header" as part of receiving a bid request to our adapter.  In OpenRTB for example the synced cookie ID is stored by a publisher or exchange and then pushed to us as a field in the bid request.

patmmccann commented:
Browsers will call your endpoint, not publisher technology, so you
will have the http header directly bc the browser is calling your
endpoint directly.

On Thu, Dec 17, 2020 at 12:52 PM Greg Loeffelholz
<notifications@github.com> wrote:
>
> @bretg - We are just releasing an adapter for now, not a prebid server. So both product and engineering on our side are still not understanding how our standard 3rd party cookie is going to show up in the bid requests that are sent to us. The endpoint we provided will just give the publisher our Cookie ID. I understand that will store in 3rd party cookie space. However, when our adapater is placed by a publisher we should then receive bid requests. The publisher initiates those through GAM for example and our adapter receives the request. We are not clear on how we will be able to see the "http header" as part of receiving a bid request to our adapter. In OpenRTB for example the synced cookie ID is stored by a publisher or exchange and then pushed to us as a field in the bid request.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

bretg commented:
@patmmccann is correct. This is just standard 3rd party cookie stuff.

1) Publisher places your adapter in page
2) User hits the page
3) Prebid.js calls the auction. For the first auction, your endpoint will most likely be called without your 3rd party cookie
4) After the first auction Prebid.js calls the usersync endpoints
5) Your endpoint will respond with image or iframe html, setting a 3rd party cookie in the meantime, pretty much only the Chrome browser though. Safari and Firefox have them off by default.
6) User hits another page
7) Prebid.js calls the new auction. This time your endpoint will receive a 'cookie' HTTP header
8) You read the HTTP header and smile
9) Don't get too used to it. 3rd party cookies are going away in about a year.
